src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php

```diff
@@ -182,6 +167,130 @@ private function getEventFromTypeDeclaration(ContainerBuilder $container, string
+            foreach ($events as $event) {
+                if (!isset($event['event']) && $container->getDefinition($id)->hasTag('kernel.event_subscriber')) {
+                    continue;
+                }
+
+                $this->guessEventAndMethod($event, $container, $id, $aliases);
+
+                if (isset($event['before']) || isset($event['after'])) {
+                    if (isset($event['before']) && isset($event['after'])) {
+                        throw new InvalidArgumentException(sprintf('Bad listener definition for "%s": cannot set "after" and "before" at the same time.', $id));
+                    }
+
+                    if (($event['priority'] ?? 0) !== 0 && (isset($event['before']) || isset($event['after']))) {
+                        throw new InvalidArgumentException(sprintf('Bad listener definition for "%s": cannot set "priority" and "before" at the same time.', $id));
+                    }
+
+                    $listeners[$event['event']]["{$id}::{$event['method']}"] = [
+                        'before' => $this->normalizeBeforeAfter($event['before'] ?? null, $aliases),
+                        'after' => $this->normalizeBeforeAfter($event['after'] ?? null, $aliases),
+                    ];
+                } else {
+                    $listeners[$event['event']]["{$id}::{$event['method']}"] = ['priority' => $event['priority'] ?? 0];
+                }
+            }
+        }
+
+        foreach ($listeners as $eventName => $listenersForEvent) {
+            foreach ($listenersForEvent as $listenerName => $listenerDefinition) {
+                if (isset($listenerDefinition['before']) || isset($listenerDefinition['after'])) {
+                    if ($listenerDefinition['before'] && !isset($listenersForEvent[$listenerDefinition['before']])) {
+                        throw new InvalidArgumentException(sprintf('Given "before" "%s" for listener "%s" does not listen the same event.', $listenerDefinition['before'], $listenerName));
+                    }
+
+                    if ($listenerDefinition['after'] && !isset($listenersForEvent[$listenerDefinition['after']])) {
+                        throw new InvalidArgumentException(sprintf('Given "after" "%s" for listener "%s" does not listen the same event.', $listenerDefinition['after'], $listenerName));
+                    }
+
+                    $this->updateListenerPriority(
+                        $container,
+                        $eventName,
+                        $listenerName,
+                        $this->computePriority($listenersForEvent, $listenerName)
+                    );
+                }
+            }
+        }
+    }
+
+    private function guessEventAndMethod(array &$eventDefinition, ContainerBuilder $container, string $serviceId, array $aliases): void
+    {
+        if (!isset($eventDefinition['event'])) {
+            $eventDefinition['method'] ??= '__invoke';
```

Comment:
Usually, classes have only 1 method that listen to a given event. Instead of assuming no-method is `__invoke`, the behavior should be to find the method of the class that have a listener for this event. Or throw an exception if there are more than 1.
This would solve en "pseudo-closure" issue by just setting the class name in the `AsEventListener` attribute.

Reply:
Yes, I'm perfectly fine with this, but this code was [extracted from the original compiler pass](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php#L74-L81), 
So changing this would change the main behavior of how the listener method is resolved.

By the way, I've just noticed what I thing is a bug while playing around with this:
```php
#[AsEventListener(event: KernelEvents::REQUEST, priority: 5)]
class MyListenerNotInvokable
{
    public function someMethod(RequestEvent $event): {}
}
```
It registers a	listener on `MyListenerNotInvokable::onKernelRequest()` and then an error is dispatched at runtime: `Error: Call to undefined method App\EventListener\MyListenerNotInvokable::onKernelRequest()`. Whereas the correct behavior would be to resolve the listener method to the only public method in the class (or throw an exception if several public methods exist)

I think I'll fix this in another PR before finishing the work here

Reply:
I understand that you've extracted the existing code, but these are 2 different things.
- the existing code finds the method to call when declaring a listener by analyzing the class (or with a fallback to a default method name). I think it's dangerous to use the public method if it's unique. There's a risk that the code will break if you have to add a new public method.
- the new code must be based on existing listeners only. It's then no longer a question of detecting public methods, but only of searching among the methods attached to a listener on the event concerned.

Reply:
yes I totally agree but don't you think that registering a listener on `MyListenerNotInvokable::onKernelRequest()` (see example above) which does not exist is a mistake? which by the way throws a runtime exception as soon as the event is dispatched although we could have detected the problem at boot time. See [this code](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php#L85-L95) which seems buggy: if the computed method name does not exist, and if `__invoke` does not exist, then we still keep the wrong method name.

What I have in mind is to remove the automatic fallback on `__invoke` [here](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php#L79) and [here](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php#L93) and replace it with a more fine-grained method resolving (would be done in another PR)

> Or throw an exception if there is not exactly 1.

this would be done exclusively for listeners declaring before/after, when resolving their priority: if not enough information is provided (ie: only class/serviceId is provided and the given class contains multiple method listening to the same event) then we'd throw an exception because before/after declaration is ambiguous

maybe these are two different things, but the latter will benefit from the former because I've reused all the code in `RegisterListenersPass::process()` to iterate on listeners definitions

WDYT?

Reply:
> This is indeed a bug. Would you mind opening a PR or an issue?

yep I'll open a PR soon

> I'm still not convinced that it can be the same mechanism for both. For the before/after feature, we know the actual listeners of the given event, which restrict the list of possible method names. I would not modify in this PR the way the listener method are guessed when they are register.

I've almost finished the work on the compiler pass with the definition you suggested:
```php
class AsEventListener
{
    public function __construct(
        public ?string $event = null,
        public ?string $method = null,
        public int $priority = 0,
        public ?string $dispatcher = null,
        /** @var string|array{class?: class-string, service?: string, method?: string} */
        public string|array|null $before = null,
        /** @var string|array{class?: class-string, service?: string, method?: string} */
        public string|array|null $after = null,
    ) {
    }
}
```
I'll push the code soon, the discussion will be easier then :)

Reply:
just a word about this:

> I would not modify in this PR the way the listener method are guessed when they are register.

even if I've refactored the compiler pass, I've touched nothing about the logic and how the way listener methods are guessed.
All I do is iterating on the listener definitions, and recomputing the priority of the listeners which have before/after properties